docs(hep): NIC Hotplug [skip ci]

wheatdog opened the pull request:
<!-- 
!IMPORTANT!
Please do not create a Pull Request without creating an issue first.
-->

#### Problem:
<!-- Explain the problem you aim to resolve in this PR. -->

Please refer to the HEP.

#### Solution:
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

Please refer to the HEP.

#### Related Issue(s):
<!--
Use `Issue #<issue number>` or `Issue harvester/harvester#<issue number>` or `Issue (paste link of issue)`. DON'T use `Fixes #<issue number>` or `Fixes (paste link of issue)`, as it will automatically close the linked issue when the PR is merged.
-->

#7042 

#### Test plan:
<!-- Describe the test plan by steps. -->

Please refer to the HEP.

#### Additional documentation or context

Please refer to the HEP.

wheatdog committed:
docs(hep): NIC Hotplug [skip ci]

Signed-off-by: Tim Liou <tim.liou@suse.com>

rrajendran17 commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+### Goals
+
+- Enable users to hot-plug a `virtio` interface connected via `bridge` or `SR-IOV` binding to a running VM without downtime.
+- Enable users to hot-unplug an interface connected via `bridge` binding from a running VM without downtime.
```
Since kubevirt supports model virtio and bridge type, both for hotplug and hotunplug scenario you can mention model virtio and type bridge  so that both statements will be in sync

wheatdog replied:
I've further tested hot-plug / hot-unplug using different models like e1000, e1000e, ne2k_pci, pcnet, rtl8139. To my surprise that most of them actually behave correctly within the guest OS. I can see them popped up or disappeared by inspecting the output of `ip a`. This could be the guest OS of my choice compensates the limitation of legacy NIC models. 

Given that `virtio` is often the most recommended model for modern guests, we can limit the goal to `virtio` model only for both hot-plug and hot-unplug. This could also reduce the QA efforts.

rrajendran17 commented on enhancements/20251009-nic-hotplug.md:
````diff
@@ -0,0 +1,133 @@
+        enabled: true
+      dynamicNetworksController:
+        enabled: true
+```
````
We need to also enable the featureGate for HotplugNICs in kubevirt CRD.
Add HotplugNICs under featureGates of kubectl edit kubevirt.kubevirt.io/kubevirt -n harvester-system.

https://github.com/harvester/harvester/blob/14ee9af0c8034d7a133e82254dc183803a63395f/deploy/charts/harvester/values.yaml#L71

rrajendran17 commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Regression tests
+
+Since this enhancement migrates the Multus CNI from a thin to a thick plugin architecture, all existing network functionalities, including the storage network, must be thoroughly tested to prevent regressions.
```
I remember the LH pods like instance manager and backing image will also be not restarted using hotpluggable feature when storage networks is configured

rrajendran17 commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+
+## Note
+
+The KubeVirt documentation[^1] specifies a maximum number of interfaces that can be hot-plugged. This limitation should be clearly communicated in the Harvester user documentation.
```
For normal virtio interfaces, we can support upto 256 maximum as per my observation.
Since multi function is on and with this, multiple interfaces can use single PCI slot, and in our case upto 8 interfaces are using 1 PCI slot, With this we will be able to extend up to 32 * 8 = 256 interfaces as best case.

I think it will be good if we can test this and document the best case limit with hotplugsnics as well.

wheatdog replied:
> Kubevirt reserves resources for 4 interface to allow later hotplug operations. The actual maximum amount of available resources depends on the machine type (e.g. q35 adds another PCI slot). For more information on maximum limits, see [libvirt documentation](https://libvirt.org/pci-hotplug.html).

The quote above is from https://kubevirt.io/user-guide/network/hotplug_interfaces/#virtio-limitations. Turns out the actual behavior is complex. https://github.com/kubevirt/kubevirt/pull/14754 implemented a new way to determine the number of reserved PCI ports.

Following is my test result on Harvester

1 vCPU 2Gi Memory: 4 free PCI slots
```
$ lspci -t -v
-[0000:00]-+-00.0  Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
           +-01.0  Device 1234:1111
           +-02.0-[01]----00.0  Red Hat, Inc. Virtio 1.0 network device
           +-02.1-[02]--
           +-02.2-[03]--
           +-02.3-[04]--
           +-02.4-[05]----00.0  Red Hat, Inc. Virtio 1.0 SCSI
           +-02.5-[06]----00.0  Red Hat, Inc. Virtio 1.0 console
           +-02.6-[07]----00.0  Red Hat, Inc. Virtio 1.0 block device
           +-02.7-[08]----00.0  Red Hat, Inc. Virtio 1.0 block device
           +-03.0-[09]----00.0  Red Hat, Inc. Virtio 1.0 memory balloon
           +-03.1-[0a]--
           +-1f.0  Intel Corporation 82801IB (ICH9) LPC Interface Controller
           +-1f.2  Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode]
           \-1f.3  Intel Corporation 82801I (ICH9 Family) SMBus Controller
``` 

1 vCPU 3Gi memory: 10 free PCI slots
```
$ lspci -t -v
-[0000:00]-+-00.0  Intel Corporation 82G33/G31/P35/P31 Express DRAM Controller
           +-01.0  Device 1234:1111
           +-02.0-[01]----00.0  Red Hat, Inc. Virtio 1.0 network device
           +-02.1-[02]--
           +-02.2-[03]--
           +-02.3-[04]--
           +-02.4-[05]--
           +-02.5-[06]--
           +-02.6-[07]--
           +-02.7-[08]--
           +-03.0-[09]--
           +-03.1-[0a]--
           +-03.2-[0b]----00.0  Red Hat, Inc. Virtio 1.0 SCSI
           +-03.3-[0c]----00.0  Red Hat, Inc. Virtio 1.0 console
           +-03.4-[0d]----00.0  Red Hat, Inc. Virtio 1.0 block device
           +-03.5-[0e]----00.0  Red Hat, Inc. Virtio 1.0 block device
           +-03.6-[0f]----00.0  Red Hat, Inc. Virtio 1.0 memory balloon
           +-03.7-[10]--
           +-1f.0  Intel Corporation 82801IB (ICH9) LPC Interface Controller
           +-1f.2  Intel Corporation 82801IR/IO/IH (ICH9R/DO/DH) 6 port SATA Controller [AHCI mode]
           \-1f.3  Intel Corporation 82801I (ICH9 Family) SMBus Controller
```

I'll try to add this to our document

rrajendran17 commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+
+### Upgrade strategy
+
+The NIC hot-plug feature must be compatible with VMs created prior to the upgrade. Additionally, the migration of Multus CNI from the thin to the thick plugin architecture must be verified to ensure a seamless Harvester upgrade process.
```
May be we can add cases like what can be affected by upgrade.
I think since creation/addition of interfaces is done after the node and pods are running the newer version, ideally there should not be any issues with hotplugnics.

rrajendran17 commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+
+The KubeVirt documentation[^1] specifies a maximum number of interfaces that can be hot-plugged. This limitation should be clearly communicated in the Harvester user documentation.
+
+A `RestartRequired` condition has been observed on the `VirtualMachine` resource after a NIC hot-plug operation is followed by a VM reboot. This appears to be incorrect behavior, and we will coordinate with the KubeVirt community to investigate and resolve the issue.
```
I think this can become a blocker later on where network config changes on VM will require a reboot which defeats the purpose of hotplugnics.

wheatdog replied:
I'll follow up closely with KubeVirt via [this issue](https://github.com/kubevirt/kubevirt/issues/15881)

wheatdog replied:
This issue doesn't occur with vanilla KubeVirt. The problem stemmed from the VMI mutating webhook in Harvester, where mac addresses were set in the VMI spec but weren't backported to the VM spec. This behavior (https://github.com/harvester/harvester/pull/8339) is introduced to fix another issue.

I'm trying to come up with a solution for it.

wheatdog commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+
+Since this enhancement migrates the Multus CNI from a thin to a thick plugin architecture, all existing network functionalities, including the storage network, must be thoroughly tested to prevent regressions.
+
+#### Functional tests
```
As suggested by @ihcsim during meeting, we could try hot-plug and hot-unplug multiple times and observe the behavior.

ibrokethecloud commented:
We should call out explicitly that in the current implementation we will not be supporting `kubeovn` based VM networks as they use a custom binding to allow dhcp capability to work correctly.

From an implementation perspective, we should have a few more api calls
* `isHotPluggingPossible` which checks if VM is live migrate-able based on conditions on VM status, and if we have not already hit the upper limit on allowable hot plug count for nics.
* `VMNetworkList` which returns bridge vlan based VM Networks that can be used to create hot plug nics for the VM by filter VMNetworks from the namespace of the VM
*

mingshuoqiu commented on enhancements/20251009-nic-hotplug.md:
```diff
@@ -0,0 +1,133 @@
+3.  Verify that attempts to add a NIC with any other binding method are blocked.
+4.  Verify that a NIC with the `bridge` binding method can be removed from a running VM.
+5.  Verify that attempts to remove a NIC with any other binding method are blocked.
+
```
Maybe we can add corner test case.
 -  Add test for invalid cases (unsupported binding, removing non-existent NIC).

What is the expected VM condition or event for confirming success?

w13915984028 commented:
Thanks for the PR,  connect with main issue https://github.com/harvester/harvester/issues/8283, and also a few comments:

(1) For guest cluster VMs(with label `harvesterhci.io/creator: docker-machine-driver-harvester`), they should/might need to be protected to hot-plug nic before the integration with Rancher Manager is done.

(2) For a running VM, are there identifications to show which NICs were ever hot-pluged? Is it allowed to un-plug any existing NICs (bridge or sriov type) or the original initial NIC is protected?

(3) Cross check with managed-dhcp https://docs.harvesterhci.io/v1.6/advanced/addons/managed-dhcp#limitations

wheatdog committed:
fix: revise the HEP

- limit the scope to virtio model and bridge binding method
- use migration-based method
- add new APIs to listHotpluggableVmNetworks and listHotunpluggableNics
- allow MAC addresses be specified during hot-plugging
- adjust the MAC addresses preserving strategy to prevent unpexected RestartRequired
- refine the test plan

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
chores: update list indentation

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
Refine HEP after FE sync up

1. **List Hotpluggable VM Networks** will be handled by FE
2. Explicty define the Hot-pluggable VM Networks
3. Rename api actions

Signed-off-by: Tim Liou <tim.liou@suse.com>

wheatdog committed:
Protection for guest cluster VMs

Signed-off-by: Tim Liou <tim.liou@suse.com>

w13915984028 reviewed (approved):
LGTM, thanks.
